sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,1345 @@
+        function will be normalized so that the push loss has weight 1 and
+        hence the pull loss has weight ``(1 - μ)/μ``.
+
+    impostor_store : str ['auto'|'list'|'sparse'], optional
+        list :
+            Three lists will be used to store the indices of reference
+            samples, the indices of their impostors and the (squared)
+            distances between the (sample, impostor) pairs.
+
+        sparse :
+            A sparse indicator matrix will be used to store the (sample,
+            impostor) pairs. The (squared) distances to the impostors will be
+            computed twice (once to determine the impostors and once to be
+            stored), but this option tends to be faster than 'list' as the
+            size of the data set increases.
+
+        auto :
+            Will attempt to decide the most appropriate choice of data
+            structure based on the values passed to :meth:`fit`.
+
+    max_iter : int, optional (default=50)
+        Maximum number of iterations in the optimization.
+
+    tol : float, optional (default=1e-5)
+        Convergence tolerance for the optimization.
+
+    callback : callable, optional (default=None)
+        If not None, this function is called after every iteration of the
+        optimizer, taking as arguments the current solution (transformation)
+        and the number of iterations. This might be useful in case one wants
+        to examine or store the transformation found after each iteration.
+
+    store_opt_result : bool, optional (default=False)
+        If True, the :class:`scipy.optimize.OptimizeResult` object returned by
+        :meth:`minimize` of `scipy.optimize` will be stored as attribute
+        ``opt_result_``.
+
+    verbose : int, optional (default=0)
+        If 0, no progress messages will be printed.
+        If 1, progress messages will be printed to stdout.
+        If > 1, progress messages will be printed and the ``iprint``
+        parameter of :meth:`_minimize_lbfgsb` of `scipy.optimize` will be set
+        to ``verbose - 2``.
+
+    random_state : int or numpy.RandomState or None, optional (default=None)
+        A pseudo random number generator object or a seed for it if int.
+
+    n_jobs : int, optional (default=1)
+        The number of parallel jobs to run for neighbors search.
+        If ``-1``, then the number of jobs is set to the number of CPU cores.
+        Doesn't affect :meth:`fit` method.
+
+    Attributes
+    ----------
+    transformation_ : array, shape (n_features_out, n_features)
+        The linear transformation learned during fitting.
+
+    n_neighbors_ : int
+        The provided ``n_neighbors`` is decreased if it is greater than or
+        equal to  min(number of elements in each class).
+
+    n_iter_ : int
+        Counts the number of iterations performed by the optimizer.
+
+    opt_result_ : scipy.optimize.OptimizeResult (optional)
```

Comment:
When is this practically useful to the user? Is it clear to the user how to interpret this data structure? Perhaps it's too much information?

Reply:
I could copy the description from scipy.optimize.OptimizeResult. The information that I find useful is the the termination reason and the hessian approximation. @agramfort also suggested this might be too much, therefore I opted for the flag which is set to `False` by default. You think I should remove it?

Reply:
Should termination reason result in a ConvergenceWarning rather than expecting the user to know how to inspect this object?

An option to store it only seems relevant if there is a memory concern about storing it, not if there are questions about its usefulness for most users.

Reply:
Yes, the option to store it is mainly because of the memory concern (the hessian approximation). And yes, I should print a `ConvergenceWarning` based on the `opt_result`.

Reply:
I'm not sure how you can get the object returned by scipy's `minimize()` with a callback? It contains much more than the transformation matrix.

Reply:
They can always compute say the loss/gradient with the transformation matrix, but yes, any object wanted to be monitored needs to be built from the transformation matrix, so indeed it's not that easy